File: python/fit/bump_hunt_evaluator.py
#!/usr/bin/python

###############
#   Imports   #
###############

import argparse
import sys
import ROOT as r
from rootpy.io import root_open
import BumpHunter as bh
import numpy as np

############
#   Main   #
############

def main() : 

    # Parse all command line arguments using the argparse module
    parser = argparse.ArgumentParser(description='')
    parser.add_argument("-i", "--input",  help="ROOT file containing toy histograms.")
    parser.add_argument("-n", "--number", help="Number of histograms to use in the evaluation.")
    parser.add_argument("-o", "--order",  help="Polynomial order used to model background.")
    parser.add_argument("-w", "--window", help="Window size.")
    parser.add_argument("-s", "--start",  help="Starting mass value.")
    parser.add_argument("-e", "--end",    help="End mass value.")
    args = parser.parse_args()

    if args.input is None : 
        print '[ Evaluator ]: A ROOT file needs to be specified!'
        sys.exit(2)

    # Open the ROOT file
    root_file = root_open(args.input)

    # Retrieve all the toy histograms from the file
    histos = list(root_file.objects(r.TH1))
    n_histos = len(histos)
    if args.number is not None : n_histos = int(args.number)
    print "[ Evaluator ] Using " + str(n_histos) + " histograms in evaluation of fitter."

    poly_order = 3
    if args.order is not None : poly_order = int(args.order)
    print "[ Evaluator ] Background will be fit using a " + str(poly_order) + " polynomial."

    window_size = 0.010
    if args.window is not None : window_size = float(args.window)
    print "[ Evaluator ] Using a window size " + str(window_size) 

    window_start = 0.03
    if args.start is not None : window_start = float(args.start)

    window_end = 0.04
    if args.end is not None : window_end = float(args.end)

    if (window_end - window_start) < window_size : 
        print "[ Evaluator ] Fit range is smaller than window size!"
        sys.exit(2)

    output_file_name = "eval_histos" + str(n_histos) + "_poly" + str(poly_order) + "_window" + str(window_size)
    output_file = r.TFile(output_file_name + ".root", "recreate")
    tree = r.TTree("evaluation", "evaluation")
    sig = np.zeros(1, dtype=float)
    sig_error = np.zeros(1, dtype=float)
    sig_pull = np.zeros(1, dtype=float)
    tree.Branch('signal', sig, 'signal/D')
    tree.Branch('signal_error', sig_error, 'signal_error/D')
    tree.Branch('sig_pull', sig_pull, 'sig_pull/D')

    # Prepare the fitter
    bump_hunter = bh.BumpHunter(poly_order)
    bump_hunter.set_mass_window_size(window_size)

    for histo_n in range(0, n_histos) : 
        results = bump_hunter.fit(histos[histo_n], window_start, window_end, 0.0005)
        final_params = results[0].floatParsFinal()
        sig[0] = final_params[final_params.index("nsig")].getVal()
        sig_error[0] = final_params[final_params.index("nsig")].getError()
        sig_pull[0] = sig[0]/sig_error[0]
        tree.Fill()

    output_file.Write()
    output_file.Close()

if __name__ == "__main__" : 
    main()

File: src/fit/bump_hunt_evaluator.cxx
/**
 *
 */

//------------------//
//--- C++ StdLib ---//
//------------------//
#include <cstdlib>
#include <iostream>
#include <getopt.h>
#include <string>

#include <TFile.h>
#include <TH1.h>
#include <RooFitResult.h>

#include <RootFileReader.h>
#include <BumpHunter.h>
#include <FlatTupleMaker.h>

using namespace std;

int main(int argc, char **argv) { 

    // Name of file to process
    string file_name;

    // Default number of histograms to process
    int hist_count = 10;

    // Default polynomial order to use to model the background
    int poly_order = 3;

    // Default mass window size to use when fitting
    double window_size = 0.020; 

    // Default start position of the mass window
    double window_start = 0.03;

    // Maximum position of the end of the mass window
    double window_end = 0.050;

    // Only fit the background
    bool bkg_only = false; 

    // Log fit results
    bool log_fit = false; 

    // Parse all the command line arguments.  If there are no valid command
    // line arguments passed, print the usage and exit the application
    static struct option long_options[] = {
        {"bkg_only",   required_argument, 0, 'b'},
        {"file_name",  required_argument, 0, 'i'},
        {"number",     required_argument, 0, 'n'},
        {"order",      required_argument, 0, 'o'},
        {"window",     required_argument, 0, 'w'},
        {"start",      required_argument, 0, 's'},
        {"end",        required_argument, 0, 'e'},
        {"log",        no_argument,       0, 'l'},
        {"help",       no_argument,       0, 'h'},
        {0, 0, 0, 0}
    };
    
    int option_index = 0;
    int option_char; 
    while ((option_char = getopt_long(argc, argv, "bi:n:o:w:s:e:lh", long_options, &option_index)) != -1) {

        switch(option_char) {
            case 'b': 
                bkg_only = true;
                break; 
            case 'i': 
                file_name = optarg;
                break;
            case 'n':
                hist_count = atoi(optarg);
                break;
            case 'o': 
                poly_order = atoi(optarg);
                break;
            case 'w':
                window_size = atof(optarg);
                break;
            case 's':
                window_start = atof(optarg);
                break;
            case 'e':
                window_end = atof(optarg);
                break;
            case 'l': 
                log_fit = true;
                break; 
            case 'h':
                return EXIT_SUCCESS; 
            default: 
                return EXIT_FAILURE;
        }
    }

    // Make sure a file was specified by the user.  If not, warn the user and 
    // exit the application.
    if (file_name.empty()) { 
        cerr << "[ EVALUATOR ]: Please specify a file to process." << endl;
        cerr << "[ EVALUATOR ]: Use --help for usage." << endl;
        return EXIT_FAILURE;
    }

    // Open the ROOT file.  If any problems are encountered, warn the user
    // and exit the application.
    TFile* file = new TFile(file_name.c_str());
    if (file->IsZombie()) { 
        cerr << "[ EVALUATOR ]: Failed to open file " << file_name.c_str(); 
        return EXIT_FAILURE;
    }

    // Create a ROOT file reader and parse the file
    RootFileReader* reader = new RootFileReader(); 
    reader->parseFile(file);

    // Create a new Bump Hunter instance and set the given properties.
    BumpHunter* bump_hunter = new BumpHunter(poly_order);
    bump_hunter->setWindowSize(window_size);
    if (bkg_only) bump_hunter->fitBkgOnly(); 
    if (log_fit) bump_hunter->writeResults();  

    // Build the string that will be used for the results file name
    string output_file = "order" + to_string(poly_order) + "_window" + to_string(int(window_size*1000)) + "mev"; 
    if (bkg_only) output_file += "_bkg_fit_result.root"; 
    else output_file += "_sig_fit_result.root"; 

    // Create a new flat ntuple and define the variables it will encapsulate.
    FlatTupleMaker* tuple = new FlatTupleMaker(output_file, "results"); 

    tuple->addVariable("ap_mass"); 
    tuple->addVariable("sig_yield"); 
    tuple->addVariable("sig_yield_error"); 
    tuple->addVariable("bkg_yield"); 
    tuple->addVariable("bkg_yield_error"); 
    tuple->addVariable("nll");
    tuple->addVariable("invalid_nll"); 
    tuple->addVariable("minuit_status");
    tuple->addVariable("edm"); 

    int hist_counter = 0; 
    for (auto& hist : reader->get1DHistograms()) {
        
        // Process only the maximum number of histograms present. 
        if (hist_counter == hist_count) break;

        map<double, RooFitResult*> results = bump_hunter->fit(hist, window_start, window_end, 0.001);
       
        for (auto& result : results) { 
           
            RooFitResult* fit_result = result.second; 
          
            // Retrieve all of the result of interest. 
            double bkg_yield = ((RooRealVar*) fit_result->floatParsFinal().find("bkg yield"))->getVal();
            double bkg_yield_error = ((RooRealVar*) fit_result->floatParsFinal().find("bkg yield"))->getError();
            double nll = fit_result->minNll();
            double invalid_nll = fit_result->numInvalidNLL();
            double minuit_status = fit_result->status();
            double edm = fit_result->edm(); 

            // Set the values of the results that will be written to the ntuple.
            tuple->setVariableValue("ap_mass", result.first);  
            tuple->setVariableValue("bkg_yield", bkg_yield);  
            tuple->setVariableValue("bkg_yield_error", bkg_yield_error);
            tuple->setVariableValue("nll", nll); 
            tuple->setVariableValue("invalid_nll", invalid_nll); 
            tuple->setVariableValue("minuit_status", minuit_status);
            tuple->setVariableValue("edm", edm);  

            // If this isn't a background only fit evaluation, skip it.
            if (!bkg_only) {
                double signal_yield = ((RooRealVar*) fit_result->floatParsFinal().find("signal yield"))->getVal();
                double signal_yield_error = ((RooRealVar*) fit_result->floatParsFinal().find("signal yield"))->getError();
                tuple->setVariableValue("sig_yield", signal_yield);  
                tuple->setVariableValue("sig_yield_error", signal_yield_error);
            }

              
            // Fill the ntuple
            tuple->fill(); 

            // Delete the fit results from memory
            delete fit_result;  
        }

        // Clear the map of results
        results.clear(); 

        // Increment the histogram counter 
        hist_counter++;
    }

    // Close the ntuple
    tuple->close(); 

    delete bump_hunter; 
    delete reader; 
    delete file;
}
